src/goto-instrument/contracts/utils.cpp

```diff
@@ -70,16 +105,68 @@ void havoc_if_validt::append_scalar_havoc_code_for_expr(
 void havoc_assigns_targetst::append_havoc_code_for_expr(
   const source_locationt location,
   const exprt &expr,
-  goto_programt &dest) const
+  goto_programt &dest)
 {
   if(expr.id() == ID_pointer_object)
   {
+    // pointer_object is still used internally to support malloc/free
     append_object_havoc_code_for_expr(
       location, to_pointer_object_expr(expr).pointer(), dest);
     return;
   }
-
-  havoc_utilst::append_havoc_code_for_expr(location, expr, dest);
+  else if(can_cast_expr<side_effect_expr_function_callt>(expr))
+  {
+    const auto &funcall = to_side_effect_expr_function_call(expr);
+    // type-checking ensures the function expression is necessarily a symbol
+    const auto &ident = to_symbol_expr(funcall.function()).get_identifier();
+    if(ident == CPROVER_PREFIX "object_whole")
+    {
+      append_object_havoc_code_for_expr(
+        location, funcall.arguments().at(0), dest);
+    }
+    else if(ident == CPROVER_PREFIX "object_from")
+    {
+      const auto ptr = typecast_exprt::conditional_cast(
+        funcall.arguments().at(0), pointer_type(char_type()));
+      const auto size = typecast_exprt::conditional_cast(
+        minus_exprt{
+          typecast_exprt::conditional_cast(
+            object_size(ptr), signed_size_type()),
+          pointer_offset(ptr)},
+        size_type());
```

Comment:
1. The changes I'll propose will help you get rid of `#include <util/c_types.h>` that you added (and getting rid of that is desirable).
2. Your `minus_exprt` is now (as of f8cc13a88) type inconsistent, because `pointer_offset` is now unsigned.
3. Please create a temporary for `object_size(ptr)` and use the type of the resulting expression throughout. Might look like this:
```
exprt obj_size = object_size(ptr);
minus_exprt size{obj_size, typecast_exprt::conditional_cast(pointer_offset(ptr), obj_size.type())};
```

Reply:
good catch thank you

Reply:
getting rid of `#include <util/c_types.h>` we lose access to `pointer_type`, `char_type` and `size_type` which are needed here